---
title: "tetranucleotide frequencies"
author: "Alison Baylay"
date: "30 September 2016"
output: html_document
---

```{r, message=FALSE}
library(Biostrings)
library(gplots)
library(RColorBrewer)
library(ggplot2)
library(reshape2)
library(plyr)
```

## Get fasta data
```{r}
Tn004.fasta <- readDNAStringSet('Tn004_S1_L001_redundans/cdhit1000.fa')
```

## Dinucleotide frequency clustering
```{r dinucleotide_frequency}
tet <- oligonucleotideFrequency(Tn004.fasta, 2, as.prob = T)

rownames(tet) <- names(Tn004.fasta)

cl <- hclust(dist(tet), method='ward.D2')
plot(as.dendrogram(cl), leaflab='none')
```

```{r heatmap, fig.height=16, fig.width=8}
#hmcol <- rev(colorRampPalette(brewer.pal(9, "YlGnBu"))(255))
#lab.col <- colorRampPalette(brewer.pal(9, 'RdPu'))(255)
#scaled.sizes <- trunc(width(Tn004.2kb)/max(width(Tn004.2kb))*255)

#heatmap.2(tet, dendrogram = 'row', Rowv = as.dendrogram(cl), trace = 'none', labRow = '', scale='row', col=hmcol, lhei = c(0.5, 4), RowSideColors = lab.col[scaled.sizes])
```

PCA plot - how well does clustering match up with this?
```{r pca}
cl.split <- cutree(cl, 5)
table(cl.split)

tet.pca <- prcomp(tet, center = T, scale. = T)
plot(tet.pca, type='l')

scores <- data.frame(rownames(tet), tet.pca$x[,1:3], length=width(Tn004.fasta), cluster=cl.split)
qplot(x=PC1, y=PC2, data=scores, col=as.factor(cluster))
qplot(x=PC1, y=PC3, data=scores, col=as.factor(cluster))
qplot(x=PC2, y=PC3, data=scores, col=as.factor(cluster))
```

Map dinucleotide frequency data onto contig coverage and SNP rate
```{r load_varscan_data}
fullSummary <- read.delim('Varscan.CovSNPrate.summary.tab', header=T)
clusterData <- data.frame(Chrom = gsub('\\|size[0-9]+', '', names(Tn004.fasta)),
                          cluster = cl.split)

allData <- merge(subset(fullSummary, Sample=='Tn004'), clusterData, by = 'Chrom', all.x=T)
allData[is.na(allData$cluster),'cluster'] <- 0

#Coverage distributions
ggplot(allData, aes(x=medCov)) + geom_histogram(binwidth=5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 150)) + geom_freqpoly(aes(x=medCov, col=as.factor(cluster)), binwidth=5, size=2)

#SNP rate distrubutions
ggplot(allData, aes(x=SNPrate*100)) + geom_histogram(binwidth=0.5, alpha=0.2, col='black') + geom_freqpoly(aes(x=SNPrate*100, col=as.factor(cluster)), binwidth=0.5, size=2) + scale_x_continuous(limits = c(0,10))
```

```{r di_varFreq}
# Load SNP-only dataset
snpSummary <- read.delim('Tn004_S1_L001_v_cdhit.VarScanSNPs.tab', header=T)

snpSummary <- cbind(colsplit(snpSummary$Chrom, '\\|size', c("Chrom", "ContigLength")), snpSummary[2:19])

snpSummary$VarFreq <- as.numeric(gsub('%','',snpSummary$VarFreq))

snpSummary['MinorAlleleFreq'] <- ifelse(snpSummary$VarFreq <= 50, snpSummary$VarFreq, 100-snpSummary$VarFreq)

# Merge with cluster data (keep only SNPs that are observed on both strands)
allSNPs <- merge(subset(snpSummary, Strands1 == 2 & Strands2 == 2), clusterData, by = 'Chrom', all.x=T)

ggplot(allSNPs, aes(x=MinorAlleleFreq)) + geom_histogram(binwidth=1, alpha=0.2, col='black') + geom_freqpoly(aes(x=MinorAlleleFreq, col=as.factor(cluster)), binwidth=1, size=2)

```

```{r blast_results}
blastTab <- read.delim('../../blast_test.tab', header=F)
colnames(blastTab) <- c('Query', 'Hit.id', 'percentID', 'Align.length', 'mismatches', 'gaps', 'qstart', 'qend', 'sstart', 'send')

blast.ss <- subset(blastTab, Align.length > 100)

blast.hit.count <- ddply(blast.ss, 'Query', 'summarise',
                         hit.count = length(percentID),
                         av.id = mean(percentID))

scores.w.blast <- merge(scores, blast.hit.count, by.x = 'rownames.tet.', by.y='Query', all.x=TRUE)

scores.w.blast$hit.count[is.na(scores.w.blast$hit.count)] <- 0

qplot(x=PC1, y=PC2, data=scores.w.blast, col=hit.count > 0, alpha=av.id)

nohits <- names(Tn004.fasta)[!names(Tn004.fasta) %in% blast.hit.count$Query]
writeXStringSet(Tn004.fasta[nohits], file='../../test_noHit.fasta', format = 'fasta', width=60)
```

## Repeat using tetranucleotide frequency
```{r tetranucleotide_frequency}
tet <- oligonucleotideFrequency(Tn004.fasta, 4, as.prob = T)
rownames(tet) <- names(Tn004.fasta)

cl <- hclust(dist(tet), method='ward.D2')
plot(as.dendrogram(cl), leaflab='none')
```

```{r tetra_pca}
cl.split <- cutree(cl, 4)
table(cl.split)

tet.pca <- prcomp(tet, center = T, scale. = T)
plot(tet.pca, type='l')

scores <- data.frame(rownames(tet), tet.pca$x[,1:3], length=width(Tn004.fasta), cluster=cl.split)
qplot(x=PC1, y=PC2, data=scores, col=as.factor(cluster))
qplot(x=PC1, y=PC3, data=scores, col=as.factor(cluster))
qplot(x=PC2, y=PC3, data=scores, col=as.factor(cluster))
```

```{r tetra_varscan}
clusterData <- data.frame(Chrom = gsub('\\|size[0-9]+', '', names(Tn004.fasta)),
                          cluster = cl.split)

allData <- merge(subset(fullSummary, Sample=='Tn004'), clusterData, by = 'Chrom', all.x=T)
allData[is.na(allData$cluster),'cluster'] <- 0

#Coverage distributions
ggplot(allData, aes(x=medCov)) + geom_histogram(binwidth=5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 150)) + geom_freqpoly(aes(x=medCov, col=as.factor(cluster)), binwidth=5, size=2)

#SNP rate distrubutions
ggplot(allData, aes(x=SNPrate*100)) + geom_histogram(binwidth=0.5, alpha=0.2, col='black') + geom_freqpoly(aes(x=SNPrate*100, col=as.factor(cluster)), binwidth=0.5, size=2) + scale_x_continuous(limits = c(0,10))


#blast results
blast.hit.count$Query <- gsub('\\|size[0-9]*', '', blast.hit.count$Query)
allData.w.blast <- merge(allData, blast.hit.count, by.x='Chrom', by.y='Query', all.x=TRUE)
allData.w.blast$hit.count[is.na(allData.w.blast$hit.count)] <- 0

ggplot(allData.w.blast, aes(x=medCov)) + geom_histogram(binwidth=5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 150)) + geom_freqpoly(aes(x=medCov, col=hit.count>0), binwidth=5, size=2)

ggplot(allData.w.blast, aes(x=SNPrate*100)) + geom_histogram(binwidth=0.5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 10)) + geom_freqpoly(aes(x=SNPrate*100, col=hit.count>0), binwidth=0.5, size=2)

ggplot(allData.w.blast, aes(x=ContigLength)) + geom_histogram(alpha=0.2, col='black') + geom_freqpoly(aes(x=ContigLength, col=hit.count>0), size=2)

#Remove cluster 4
allData.ss <- subset(allData.w.blast, cluster!=4)
ggplot(allData.ss, aes(x=medCov)) + geom_histogram(binwidth=5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 150)) + geom_freqpoly(aes(x=medCov, col=hit.count>0), binwidth=5, size=2)

ggplot(allData.ss, aes(x=SNPrate*100)) + geom_histogram(binwidth=0.5, alpha=0.2, col='black') + scale_x_continuous(limits = c(0, 10)) + geom_freqpoly(aes(x=SNPrate*100, col=hit.count>0), binwidth=0.5, size=2)

ggplot(allData.ss, aes(x=ContigLength)) + geom_histogram(alpha=0.2, col='black') + geom_freqpoly(aes(x=ContigLength, col=hit.count>0), size=2)
```

```{r tetra_varFreq}
# Merge with cluster data (keep only SNPs that are observed on both strands)
allSNPs <- merge(subset(snpSummary, Strands1 == 2 & Strands2 == 2), clusterData, by = 'Chrom', all.x=T)

ggplot(allSNPs, aes(x=MinorAlleleFreq)) + geom_histogram(binwidth=1, alpha=0.2, col='black') + geom_freqpoly(aes(x=MinorAlleleFreq, col=as.factor(cluster)), binwidth=1, size=2)

allSNPs.w.blast <- merge(allSNPs, blast.hit.count, by.x='Chrom', by.y='Query', all.x=TRUE)
allSNPs.w.blast$hit.count[is.na(allSNPs.w.blast$hit.count)] <- 0

ggplot(allSNPs.w.blast, aes(x=MinorAlleleFreq)) + geom_histogram(binwidth=1, alpha=0.2, col='black') + geom_freqpoly(aes(x=MinorAlleleFreq, col=hit.count>0), binwidth=1, size=2)

#Remove cluster 4
allSNPs.ss <- subset(allSNPs.w.blast, cluster != 4)
ggplot(allSNPs.ss, aes(x=MinorAlleleFreq)) + geom_histogram(binwidth=1, alpha=0.2, col='black') + geom_freqpoly(aes(x=MinorAlleleFreq, col=hit.count>0), binwidth=1, size=2)
```

```{r megan_assignment}
megan.csv <- '../../test_last/Tn004_v_refseq_microbial_last_RNtoTxN.csv'

megan.data <- read.csv(megan.csv, header=F, stringsAsFactors = F)
colnames(megan.data) <- c('Chrom', 'tax')
megan.data$Chrom <- gsub('\\|size[0-9]+', '', megan.data$Chrom)

allData.w.megan <- merge(allData, megan.data, by='Chrom', all.x=TRUE)
allData.w.megan$tax[is.na(allData.w.megan$tax)] <- 'No LAST hits'

unassigned.list <- c('No LAST hits', 'Not assigned')

allData.w.megan[,'Assigned'] <- !(allData.w.megan$tax %in% unassigned.list)

allData.w.megan[,'simple.tax'] <- NA
allData.w.megan$simple.tax[allData.w.megan$tax %in% unassigned.list] <- 'Unassigned'
allData.w.megan$simple.tax[allData.w.megan$tax == 'Trichodesmium erythraeum'] <- 'Trichodesmium erythraeum'

other.oscillatoriales <- c('Arthrospira', 'Crinalium', 'Geitlerinema', 'Kamptonema', 'Leptolyngbya', 'Limnoraphis', 'Lyngbya', 'Microcoleus', 'Moorea', 'Oscillatoria', 'Planktothricoides', 'Planktothrix', 'Pseudanabaena', 'Spirulina', 'unclassified Oscillatoriales')
other.cyanos <- c('Nostocales', 'Chroococcales',  'Pleurocapsales', 'Stigonematales', 'unclassified Cyanobacteria')
undefined.cyanos <- c('Cyanobacteria', 'Oscillatoriophycideae', 'Oscillatoriales')

allData.w.megan$simple.tax[allData.w.megan$tax %in% other.oscillatoriales] <- 'Other Oscillatoriales'
allData.w.megan$simple.tax[allData.w.megan$tax %in% other.cyanos] <- 'Other Cyanobacteria'
allData.w.megan$simple.tax[allData.w.megan$tax %in% undefined.cyanos] <- 'Undefined Cyanobacteria'

undefined.bact <- c('Bacteria', 'Terrabacteria group')
other <- c('Euryarchaeota', 'cellular organisms')

allData.w.megan$simple.tax[allData.w.megan$tax %in% undefined.bact] <- 'Undefined Bacteria'
allData.w.megan$simple.tax[allData.w.megan$tax %in% other] <- 'Other organisms'
allData.w.megan$simple.tax[is.na(allData.w.megan$simple.tax)] <- 'Other Bacteria'

ggplot(allData.w.megan, aes(x=medCov)) + geom_histogram(binwidth=5, alpha=0.2, col='black') + geom_freqpoly(aes(x=medCov, col=Assigned), binwidth=5, size=2) + scale_x_continuous(limits = c(0, 150))

allData.w.megan[,'v.simple.tax'] <- sapply(allData.w.megan$simple.tax, function(x) unlist(strsplit(x, ' '))[1])
```


```{r get_outliers}
#Tn004.cluster5 = Tn004.fasta[cl.split==5]
#writeXStringSet(Tn004.cluster5, file='Tn004_odd_cluster.fasta', format = 'fasta', width=60)
```